### PR TITLE
Improve type hints for cached decorator

### PR DESCRIPTION
**Pull request by clokep**
The `cached` and `cachedList` decorators return a `Deferred` regardless of whether the underlying function is async/returns a `Deferred`. The plugin we have didn't really do this properly and I was hitting an edge-case with it. This attempts to fix this up.

### CODE REVIEW

File: tests/appservice/test_appservice.py [hunk 1]
```diff
@@ -49,93 +49,81 @@ def setUp(self) -> None:
...
         self.event.sender = "@irc_foobar:matrix.org"
         self.assertTrue(
             (
-                yield defer.ensureDeferred(
-                    self.service.is_interested_in_event(
-                        self.event.event_id, self.event, self.store
-                    )
+                yield self.service.is_interested_in_event(
+                    self.event.event_id, self.event, self.store
                 )
```
Comment (clokep) [inline tests/appservice/test_appservice.py:59]: It is already a `Deferred`, so we don't need to wrap it in `ensureDeferred` here.

File: tests/appservice/test_appservice.py [hunk 2]
```diff
@@ -49,93 +49,81 @@ def setUp(self) -> None:
     @defer.inlineCallbacks
     def test_regex_user_id_prefix_match(
         self,
-    ) -> Generator["defer.Deferred[object]", object, None]:
+    ) -> Generator["defer.Deferred[Any]", object, None]:
```
Comment (clokep) [inline tests/appservice/test_appservice.py:52]: `Deferred[object]` can't yield `Deferred[bool]`, but using `Deferred[bool]` doesn't work with `inlineCallbacks`.

File: tests/storage/test_transactions.py [hunk 3]
```diff
@@ -33,15 +33,14 @@ def test_get_set_transactions(self) -> None:
...
 
-        d = self.store.set_destination_retry_timings("example.com", 1000, 50, 100)
-        self.get_success(d)
+        self.get_success(
+            self.store.set_destination_retry_timings("example.com", 1000, 50, 100)
+        )
 
-        d = self.store.get_destination_retry_timings("example.com")
-        r = self.get_success(d)
+        r = self.get_success(self.store.get_destination_retry_timings("example.com"))
```
Comment (clokep) [inline tests/storage/test_transactions.py:43]: The `get` and `set` versions now properly have different return types, I just inlined them instead of having a `d` and a `d2`.

File: scripts-dev/mypy_synapse_plugin.py [hunk 4]
```diff
@@ -92,10 +93,35 @@ def cached_function_method_signature(ctx: MethodSigContext) -> CallableType:
...
+        isinstance(signature.ret_type, Instance)
+        and signature.ret_type.type.fullname == "twisted.internet.defer.Deferred"
+    ):
+        # If it is already a Deferred, nothing to do.
+        ret_type = signature.ret_type
+    else:
+        # If a coroutine, wrap the coroutine's return type in a Deferred.
+        if (
+            isinstance(signature.ret_type, Instance)
+            and signature.ret_type.type.fullname == "typing.Coroutine"
```
Comment (DMRobertson) [inline scripts-dev/mypy_synapse_plugin.py:107]: These cases don't cover the cached function returning a generic `Awaitable[R]` . I don't _think_ we use Awaitables that aren't Deferreds or Coroutines in Synapse's code (outside of tests, at least), so this is probably fine. We can come back to this in the future if need be.

File: scripts-dev/mypy_synapse_plugin.py [hunk 5]
```diff
@@ -92,10 +93,35 @@ def cached_function_method_signature(ctx: MethodSigContext) -> CallableType:
...
+        ):
+            ret_arg = signature.ret_type.args[2]
+
+        # Otherwise, wrap the return type in a Deferred.
+        else:
+            ret_arg = signature.ret_type
+
+        # This should be able to use ctx.api.lookup_typeinfo, but that doesn't seem
+        # to find the correct symbol.
+        sym = ctx.api.modules["twisted.internet.defer"].names.get("Deferred")  # type: ignore[attr-defined]
```
Comment (DMRobertson) [inline scripts-dev/mypy_synapse_plugin.py:117]: Which attr is unknown here---`.names`?

File: scripts-dev/mypy_synapse_plugin.py [hunk 6]
```diff
@@ -92,10 +94,35 @@ def cached_function_method_signature(ctx: MethodSigContext) -> CallableType:
...
+        ret_type = Instance(sym.node, [remove_instance_last_known_values(ret_arg)])
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply scripts-dev/mypy_synapse_plugin.py:118]: Can you leave a comment explaining what `remove_instance_last_known_values` is doing (or at least, what breaks without it)?

File: synapse/storage/databases/main/roommember.py [hunk 7]
```diff
@@ -1099,7 +1099,7 @@ async def _get_joined_hosts(
         # `get_joined_hosts` is called with the "current" state group for the
         # room, and so consecutive calls will be for consecutive state groups
         # which point to the previous state group.
-        cache = await self._get_joined_hosts_cache(room_id)  # type: ignore[misc]
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/storage/databases/main/roommember.py:1102]: Assuming this means that `reveal_type(cache)` is `JoinedHostsCache`. (What was the revealed type before?)

File: synapse/util/caches/descriptors.py [hunk 8]
```diff
@@ -232,7 +234,7 @@ def __get__(self, obj: Optional[Any], owner: Optional[Type]) -> Callable[..., An
         get_cache_key = self.cache_key_builder
 
         @functools.wraps(self.orig)
-        def _wrapped(*args: Any, **kwargs: Any) -> Any:
+        def _wrapped(*args: Any, **kwargs: Any) -> "defer.Deferred[Any]":
```
Comment (DMRobertson) [inline synapse/util/caches/descriptors.py:237]: Aside: there might be scope to give types to `self.orig`'s arguments with a `ParamSpec` here in the future.

(In fact, I wonder if Python typing is expressive enough these days to not need a mypy plugin for this at all?)

File: synapse/util/caches/descriptors.py [hunk 9]
```diff
@@ -220,7 +220,9 @@ def __init__(
         self.iterable = iterable
         self.prune_unread_entries = prune_unread_entries
 
-    def __get__(self, obj: Optional[Any], owner: Optional[Type]) -> Callable[..., Any]:
+    def __get__(
+        self, obj: Optional[Any], owner: Optional[Type]
+    ) -> Callable[..., "defer.Deferred[Any]"]:
```
Comment (DMRobertson) [inline synapse/util/caches/descriptors.py:225]: Is this the method that the plugin changes target?

**Review by DMRobertson — COMMENTED** [review]
Sorry for preemptive review.

File: scripts-dev/mypy_synapse_plugin.py (hunk 4 above)
Comment (clokep) [reply scripts-dev/mypy_synapse_plugin.py:107]: Yeah, we don't seem to do that so I didn't bother covering it. I should likely just add it while I'm here though.

File: scripts-dev/mypy_synapse_plugin.py (hunk 5 above)
Comment (clokep) [reply scripts-dev/mypy_synapse_plugin.py:117]: `.modules`.

File: synapse/util/caches/descriptors.py (hunk 8 above)
Comment (clokep) [reply synapse/util/caches/descriptors.py:237]: > (In fact, I wonder if Python typing is expressive enough these days to not need a mypy plugin for this at all?)

I also had this thought but didn't want to go down that avenue at the moment.

File: synapse/util/caches/descriptors.py (hunk 9 above)
Comment (clokep) [reply synapse/util/caches/descriptors.py:225]: Kind of? The plugin doesn't really care about any of the internals of the descriptors, but I believe this corresponds to what I'm doing.

**Comment by clokep** [comment]
> Sorry for preemptive review.

I was just waiting for CI to pass. 😄

File: synapse/storage/databases/main/roommember.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/databases/main/roommember.py:1102]: It was `Any`, it is now `synapse.storage.databases.main.roommember._JoinedHostsCache`.

File: scripts-dev/mypy_synapse_plugin.py (hunk 5 above)
Comment (DMRobertson) [reply scripts-dev/mypy_synapse_plugin.py:117]: That's mildly terrifying. But mypy plugins are uncharted waters!

Comment (clokep) [reply scripts-dev/mypy_synapse_plugin.py:123]: I expanded the comment.

File: scripts-dev/mypy_synapse_plugin.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (clokep) [reply scripts-dev/mypy_synapse_plugin.py]: I expanded the comment to say where this comes from.

**Review by DMRobertson — APPROVED** [review]
Thanks!
